### PR TITLE
Use tasklist instead of checking the lockfile.

### LINKED ISSUE #3568: Use tasklist to determine if RD is still running on Windows

### PR DESCRIPTION
**Pull request by ericpromislow**
Also some refactoring to reduce function signature size.

Fixes #3568

Signed-off-by: Eric Promislow <epromislow@suse.com>

### CODE REVIEW

File: src/go/rdctl/pkg/factoryreset/factory_reset_windows.go [hunk 1]
```diff
@@ -39,6 +41,34 @@ var (
...
+	cmd.SysProcAttr = &syscall.SysProcAttr{CreationFlags: CREATE_NO_WINDOW}
+	allOutput, err := cmd.CombinedOutput()
+	if err != nil {
+		return false, fmt.Errorf("Failed to run %q: %w", cmd, err)
+	}
+	r := csv.NewReader(strings.NewReader(string(allOutput)))
+	for {
+		record, err := r.Read()
+		if err != nil {
+			if err != io.EOF {
```
Comment (mook-as) [inline src/go/rdctl/pkg/factoryreset/factory_reset_windows.go:60]: Please use `errors.Is` instead of comparing errors directly.

File: src/go/rdctl/pkg/factoryreset/factory_reset_windows.go [hunk 2]
```diff
@@ -39,6 +41,34 @@ var (
...
+				return false, fmt.Errorf("Failed to csv-read the output for tasklist: %w", err)
+			}
+			break
+		}
+		if record[0] == "Rancher Desktop.exe" {
```
Comment (mook-as) [inline src/go/rdctl/pkg/factoryreset/factory_reset_windows.go:65]: We should check if `record` has any entries before dereferencing.

File: src/go/rdctl/pkg/shutdown/shutdown.go [hunk 3]
```diff
@@ -17,34 +17,40 @@ limitations under the License.
...
 	"github.com/rancher-sandbox/rancher-desktop/src/go/rdctl/pkg/factoryreset"
 	"github.com/sirupsen/logrus"
 )
 
-func FinishShutdown(waitForShutdown bool) error {
+type ShutdownInfo struct {
+	waitForShutdown bool
+}
+
+func NewShutdownInfo(waitForShutdown bool) *ShutdownInfo {
```
Comment (mook-as) [inline src/go/rdctl/pkg/shutdown/shutdown.go:35]: We're making a struct now so we can carry `waitForShutdown` around, right? It might be cleaner to convert `waitForAppToDieOrKillIt` to use a struct instead (since it doesn't get us anything as an API exposed to other packages).

File: src/go/rdctl/pkg/shutdown/shutdown.go [hunk 4]
```diff
@@ -54,72 +60,51 @@ func FinishShutdown(waitForShutdown bool) error {
...
+func (s *ShutdownInfo) waitForAppToDieOrKillIt(checkFunc func() (bool, error), killFunc func() error, retryCount int, retryWait int, operation string) error {
+	for iter := 0; s.waitForShutdown && iter < retryCount; iter++ {
 		if iter > 0 {
 			logrus.Debugf("checking %s showed it's still running; sleeping %d seconds\n", operation, retryWait)
 			time.Sleep(time.Duration(retryWait) * time.Second)
 		}
-		if !checkFunc() {
+		status, err := checkFunc()
+		if err != nil {
+			return fmt.Errorf("checking operation %s => error %w", operation, err)
```
Comment (mook-as) [inline src/go/rdctl/pkg/shutdown/shutdown.go:71]: Sample output:
> checking operation the app => error The file was not found.

The wording seems a bit odd; consider `while checking %s, found error: %w`.

File: src/go/rdctl/pkg/shutdown/shutdown.go [hunk 5]
```diff
@@ -54,72 +60,51 @@ func FinishShutdown(waitForShutdown bool) error {
...
-		return err
-	}
-	return nil
+	return killFunc()
 }
 
 /**
- * checkProcessX function returns true if it detects the app is still running, false otherwise
+ * checkProcessX function returns [true, nil] if it detects the app is still running, [false, X] otherwise
+ * Linux/macOS errors are always nil, Windows not so
```
Comment (mook-as) [inline src/go/rdctl/pkg/shutdown/shutdown.go:84]: So functionally, the boolean return is always redundant: the error return is either nil (it is running) or non-nil (it is no longer running)? That sounds incorrect; maybe reword this?

**Review by mook-as — CHANGES_REQUESTED** [review]

File: src/go/rdctl/pkg/factoryreset/factory_reset_windows.go (hunk 2 above)
Comment (ericpromislow) [reply src/go/rdctl/pkg/factoryreset/factory_reset_windows.go:65]: Not sure how that's going to happen, as the csv reader skips empty lines, but easy enough to do.

File: src/go/rdctl/pkg/factoryreset/factory_reset_windows.go (hunk 1 above)
Comment (ericpromislow) [reply src/go/rdctl/pkg/factoryreset/factory_reset_windows.go:60]: ok

File: src/go/rdctl/pkg/shutdown/shutdown.go (hunk 5 above)
Comment (ericpromislow) [reply src/go/rdctl/pkg/shutdown/shutdown.go:84]: Not exactly. I reworded the comment. All I wanted was an error-situation in `factoryreset.CheckProcessWindows` to be propagated.

**Comment by ericpromislow** [comment]
There's a separate issue to deal with `shutdown` and `factory-reset` for `npm run dev`

File: src/go/rdctl/pkg/shutdown/shutdown.go (hunk 3 above)
Comment (ericpromislow) [reply src/go/rdctl/pkg/shutdown/shutdown.go:35]: I moved to this because at one point I needed to make `waitForShutdown` available to one of the kill routines, and didn't want to keep pushing that value through function calls. But, sure, I can make it local to the package.

File: src/go/rdctl/pkg/factoryreset/factory_reset_windows.go [hunk 6]
```diff
@@ -39,6 +41,34 @@ var (
...
+// It does this by calling `tasklist`, the Windows answer to ps(1)
+
+func CheckProcessWindows() (bool, error) {
+	cmd := exec.Command("tasklist", "/NH", "/FI", "IMAGENAME eq Rancher Desktop.exe", "/FO", "CSV")
+	cmd.SysProcAttr = &syscall.SysProcAttr{CreationFlags: CREATE_NO_WINDOW}
+	allOutput, err := cmd.CombinedOutput()
+	if err != nil {
+		return false, fmt.Errorf("Failed to run %q: %w", cmd, err)
+	}
+	r := csv.NewReader(strings.NewReader(string(allOutput)))
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/go/rdctl/pkg/factoryreset/factory_reset_windows.go]: I did that, but I had to pull in the `bytes` library, and I'm pretty sure `string`-casting is a noop at runtime, but whatever...

**Review by mook-as — APPROVED** [review]

File: src/go/rdctl/pkg/factoryreset/factory_reset_windows.go [hunk 7]
```diff
@@ -39,6 +43,34 @@ var (
 	pEnumProcesses = pKernel32.NewProc("K32EnumProcesses")
 )
 
+// CheckProcessWindows - returns true if Rancher Desktop is still running, false if it isn't
+// along with an error condition if there's a problem detecting that.
+//
+// It does this by calling `tasklist`, the Windows answer to ps(1)
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply src/go/rdctl/pkg/factoryreset/factory_reset_windows.go:49]: I don't think an `.exe` is necessary in the comment; it is an `.exe` command, but it doesn't really matter. And the invocation works without the extension.

Also the separation of the block-comments from the code makes for easiler reading IMO
